Storage/SQL/SQLiteHistory.swift

```diff
@@ -320,24 +325,28 @@ extension SQLiteHistory: BrowserHistory {
         }
 
         let ungroupedSQL =
-        "SELECT \(TableHistory).id AS historyID, \(TableHistory).url AS url, title, guid, domain_id, domain, " +
-        "COALESCE(max(case \(TableVisits).is_local when 1 then \(TableVisits).date else 0 end), 0) AS localVisitDate, " +
-        "COALESCE(max(case \(TableVisits).is_local when 0 then \(TableVisits).date else 0 end), 0) AS remoteVisitDate, " +
-        "COALESCE(count(\(TableVisits).is_local), 0) AS visitCount " +
-        "FROM \(TableHistory) " +
-        "INNER JOIN \(TableDomains) ON \(TableDomains).id = \(TableHistory).domain_id " +
-        "INNER JOIN \(TableVisits) ON \(TableVisits).siteID = \(TableHistory).id " +
-        whereClause + " GROUP BY historyID"
+            "SELECT \(TableHistory).id AS historyID, \(TableHistory).url AS url, title, guid, domain_id, domain, " +
+            "COALESCE(max(case \(TableVisits).is_local when 1 then \(TableVisits).date else 0 end), 0) AS localVisitDate, " +
+            "COALESCE(max(case \(TableVisits).is_local when 0 then \(TableVisits).date else 0 end), 0) AS remoteVisitDate, " +
+            "COALESCE(count(\(TableVisits).is_local), 0) AS visitCount " +
+            (includeBookmarked ? ", count(\(TableBookmarks).id) AS bookmarked " : "") +
+            "FROM \(TableHistory) " +
+            (includeBookmarked ? "LEFT OUTER JOIN \(TableBookmarks) ON \(TableBookmarks).url IS \(TableHistory).url " : "") +
```

Comment:
I have my doubts about joining history directly against bookmarks here, particularly using the URL. This is already a pretty expensive query.

The obvious long-term solution is to switch to a genuine places-style system with some shared indexed ID.

Other alternatives are to do a virtual join asynchronously after showing the list (which we should also do for icons), or to very simply… track bookmark state as a counter in each history item. Bookmarking is a rare event, adding history is a cheap place to do this query, etc. etc.

Can you do some analysis about the complexity change in the `EXPLAIN QUERY PLAN` here, and how this affects timing on moderate history DBs (~5000 history items) and moderate bookmark sets (~100)?